core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```
@@ -1194,6 +1216,9 @@ object GroupMetadataManager {
       if (version > 0)
         memberStruct.set(REBALANCE_TIMEOUT_KEY, memberMetadata.rebalanceTimeoutMs)
 
+      if (version >= 3)
+        memberStruct.set(GROUP_INSTANCE_ID_KEY, memberMetadata.groupInstanceId)
```

I'm trying to think through the implications of this. We are silently discarding the instance id which means that replicas won't know about it. The member will be considered a static member until there is a coordinator change. Then it will suddenly become dynamic again and I think that would trigger this assertion: https://github.com/apache/kafka/pull/6177/files#diff-a210b6f573f800e8a562fda9d53cd148R269. I think we probably need to avoid using the static membership logic entirely until the IBP supports it.

---

That sounds reasonable. I'm not sure I'm fully following here because unless broker upgrades to latest, the group instance id should not include the join group because of automatic request downgrade.

---

What's `IBP`?

---

@guozhangwang inter broker protocol

---

We expose the new JoinGroup protocol as soon as the binary is updated. The client will begin using it. That itself is fine, but it is not safe for the broker to use the static member logic until we are sure that all brokers support it, as indicated through the IBP. Otherwise, the case I mentioned is possible. We seem to have removed the assertion I mentioned above, but I am still not sure the logic is correct. The simplest option would be to set `groupInstanceId` to `None` if the IBP is below `KAFKA_2_3_IV0`.

---

Addressed in GroupCoordinator.scala